src/balancer/src/lib.rs

```diff
@@ -0,0 +1,466 @@
+    /// Listen address for pgwire connections.
+    pgwire_listen_addr: SocketAddr,
+    /// Listen address for HTTP connections.
+    http_listen_addr: SocketAddr,
+    /// DNS resolver.
+    resolver: Resolver,
+    tls: Option<TlsCertConfig>,
+}
+
+impl BalancerConfig {
+    pub fn new(
+        build_info: &BuildInfo,
+        pgwire_listen_addr: SocketAddr,
+        http_listen_addr: SocketAddr,
+        resolver: Resolver,
+        tls: Option<TlsCertConfig>,
+    ) -> Self {
+        Self {
+            build_version: build_info.semver_version(),
+            pgwire_listen_addr,
+            http_listen_addr,
+            resolver,
+            tls,
+        }
+    }
+}
+
+/// Prometheus monitoring metrics.
+#[derive(Debug)]
+pub struct Metrics {
+    _uptime: ComputedGauge,
+}
+
+impl Metrics {
+    /// Returns a new [Metrics] instance connected to the given registry.
+    pub fn new(cfg: &BalancerConfig, registry: &MetricsRegistry) -> Self {
+        let start = Instant::now();
+        let uptime = registry.register_computed_gauge(
+            metric!(
+                name: "mz_balancer_metadata_seconds",
+                help: "server uptime, labels are build metadata",
+                const_labels: {
+                    "version" => cfg.build_version,
+                    "build_type" => if cfg!(release) { "release" } else { "debug" }
+                },
+            ),
+            move || start.elapsed().as_secs_f64(),
+        );
+        Metrics { _uptime: uptime }
+    }
+}
+
+pub struct BalancerService {
+    cfg: BalancerConfig,
+    _metrics: Metrics,
+}
+
+impl BalancerService {
+    pub fn new(cfg: BalancerConfig, metrics: Metrics) -> Self {
+        Self {
+            cfg,
+            _metrics: metrics,
+        }
+    }
+
+    pub async fn serve(self) -> Result<(), anyhow::Error> {
+        let (_pgwire_listen_handle, pgwire_stream) = listen(self.cfg.pgwire_listen_addr).await?;
+        let (_http_listen_handle, http_stream) = listen(self.cfg.http_listen_addr).await?;
+
+        let pgwire_tls = match self.cfg.tls {
+            Some(tls) => Some(TlsConfig {
+                context: tls.context()?,
+                mode: TlsMode::Require,
+            }),
+            None => None,
+        };
+
+        let resolver = Arc::new(self.cfg.resolver);
+
+        let pgwire = PgwireBalancer {
+            resolver: Arc::clone(&resolver),
+            tls: pgwire_tls,
+        };
+        let http = HttpBalancer {
+            _resolver: Arc::clone(&resolver),
+        };
+
+        let pgwire_task_handle = task::spawn(|| "pgwire_stream", async move {
+            mz_ore::server::serve(pgwire_stream, pgwire).await;
+        });
+        let http_task_handle = task::spawn(|| "http_stream", async move {
+            mz_ore::server::serve(http_stream, http).await;
+        });
+
+        // These should never complete, but complain if they ever do.
+        pgwire_task_handle.await?;
+        http_task_handle.await?;
```

Comment:
as-is the http task could finish and we'd never know, unless the pgwire task also finished. You should probably select over the two futures if we care about this, using the [select](https://docs.rs/futures/latest/futures/future/fn.select.html) or [select_all](https://docs.rs/futures/latest/futures/future/fn.select_all.html) functions might be easier than a `select!` macro in this case

Reply:
Changed to use JoinSet.